readthedocs/builds/models.py

```diff
@@ -440,11 +477,75 @@ class Build(models.Model):
 
     objects = BuildQuerySet.as_manager()
 
+    CONFIG_KEY = '__config'
+
     class Meta(object):
         ordering = ['-date']
         get_latest_by = 'date'
         index_together = [['version', 'state', 'type']]
 
+    def __init__(self, *args, **kwargs):
+        super(Build, self).__init__(*args, **kwargs)
+        self._config_changed = False
+
+    @property
+    def previous(self):
+        """
+        Returns the previous build to current one.
+
+        Matching the project and version.
+        """
+        date = self.date or timezone.now()
+        if self.project is not None and self.version is not None:
+            return (
+                Build.objects
+                .filter(
+                    project=self.project,
+                    version=self.version,
+                    date__lt=date,
+                )
+                .order_by('-date')
+                .first()
+            )
+        return None
+
+    @property
+    def config(self):
+        """
+        Get the config used for this build.
+
+        Since we are saving the config into the JSON field only when it differs
+        from the previous one, this helper returns the correct JSON used in
+        this Build object (it could be stored in this object or one of the
+        previous ones).
+        """
+        if self.CONFIG_KEY in self._config:
+            return Build.objects.get(pk=self._config[self.CONFIG_KEY])._config
+        return self._config
+
+    @config.setter
+    def config(self, value):
+        self._config = value
+        self._config_changed = True
+
+    def save(self, *args, **kwargs):  # noqa
+        """
+        Save object.
+
+        To save space on the db we only save the config if it's different
+        from the previous one.
+
+        If the config is the same, we save the pk of the object
+        that has the **real** config under the `CONFIG_KEY` key.
+        """
+        if self.pk is None or self._config_changed:
```

Comment:
So, I did this to save only when needed. I still need to test a little more, but I think we are good